README.md
SkypeTab is a program that adds tabs to Skype™ for Linux.

[Project page](http://keks-n.net/skypetab) (description, demo, repositories, etc).

Compilation & Installation
==========================

First you need environment that's able to compile 32-bit code. You also need libX11 and Qt and headers for them.

Run qmake -spec linux-g++-32 && make

Then you will be able to start SkypeTab using LD_PRELOAD=\`pwd\`/libskypetab.so skype

Note, LD_PRELOAD requires full path to the library


Coding "guidelines"
==========================
If you want to participate, please, use tabs for indention and spaces for formatting. Thank you.

debian/changelog
skypetab-ng (0.5.3-ubuntu) trusty; urgency=medium

  * Added compatibility with Skype 4.3, now building only for trusty

 -- Nikita Tsukanov <keks9n@gmail.com>  Thu, 19 Jun 2014 00:26:48 +0400

skypetab-ng (0.5.2-1) quantal; urgency=low

  * Added compatibility with Skype 4.1

 -- Nikita Tsukanov <keks9n@gmail.com>  Tue, 20 Nov 2012 14:49:02 +0400

skypetab-ng (0.5.1-1) quantal; urgency=low

  * Speed up window activation from closed state 

 -- Nikita Tsukanov <keks9n@gmail.com>  Fri, 02 Nov 2012 15:22:22 +0400

skypetab-ng (0.5.0-1) quantal; urgency=low

  * Now compatible with Skype 4.0

 -- Nikita Tsukanov <keks9n@gmail.com>  Thu, 14 Jun 2012 19:20:09 +0400

skypetab-ng (0.4.11-1) precise; urgency=low

  * Added minimize-on-close behaviour for Unity

 -- Nikita Tsukanov <keks9n@gmail.com>  Thu, 31 May 2012 22:24:54 +0400

skypetab-ng (0.4.10-1) precise; urgency=low

  * Fixed several compatibility issues (Unity, KDE)

 -- Nikita Tsukanov <keks9n@gmail.com>  Wed, 09 May 2012 20:36:19 +0400

skypetab-ng (0.4.9-1) oneiric; urgency=low

  * Fixed some focus issues
  * Added "don't close tabs with the window" option
  * Added additional event counter to Skype's icon

 -- Nikita Tsukanov <keks9n@gmail.com>  Sat, 24 Dec 2011 16:55:57 +0400

skypetab-ng (0.4.8-1) oneiric; urgency=low

  * Now it remembers it's last window position
  * Added some options that are useful for autostart
  * Added tab switching using Alt+1-9 keys

 -- Nikita Tsukanov <keks9n@gmail.com>  Thu, 22 Dec 2011 18:18:35 +0400

skypetab-ng (0.4.7-1) oneiric; urgency=high

  * Clear LD_PRELOAD to prevent libskypetab-ng from loading to processes other than skype

 -- Nikita Tsukanov <keks9n@gmail.com>  Tue, 20 Dec 2011 12:18:56 +0400

skypetab-ng (0.4.6-1) oneiric; urgency=low

  * Now chat input is focused on window activation
  * Added tabs switching via Ctrl+Tab/PgDown/PgUp

 -- Nikita Tsukanov <keks9n@gmail.com>  Mon, 19 Dec 2011 04:38:44 +0400

skypetab-ng (0.4.5-1) oneiric; urgency=low

  * Added settings dialog
  * Fixed possible segfaults

 -- Nikita Tsukanov <keks9n@gmail.com>  Sun, 18 Dec 2011 23:12:27 +0400

skypetab-ng (0.4.4-2) oneiric; urgency=low

  * Fixed some bugs and perfomance issues

 -- Nikita Tsukanov <keks9n@gmail.com>  Sat, 17 Dec 2011 00:00:11 +0400

skypetab-ng (0.4.3-1) oneiric; urgency=low

  * Now using new window embedding method. It should fix most problems with window managers

 -- Nikita Tsukanov <keks9n@gmail.com>  Thu, 15 Dec 2011 22:32:25 +0400

skypetab-ng (0.4.2-1) oneiric; urgency=low

  * Fixed some Unity compatibility issues
  * Fixed some problems with focus

 -- Nikita Tsukanov <keks9n@gmail.com>  Thu, 15 Dec 2011 16:18:06 +0400

skypetab-ng (0.4.1-7) oneiric; urgency=low

  * Removed build for amd64

 -- Nikita Tsukanov <keks9n@gmail.com>  Wed, 14 Dec 2011 12:37:40 +0400

skypetab-ng (0.4.1-1) oneiric; urgency=low

  * Initial packaging 

 -- Nikita Tsukanov <keks9n@gmail.com>  Tue, 13 Dec 2011 20:30:57 +0400


skypetab.cpp
/*
skypetab.cpp
Author:
     Nikita Tsukanov <keks9n@gmail.com>

Copyright (c) 2011 Nikita Tsukanov

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Lesser General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU Lesser General Public License for more details.

You should have received a copy of the GNU Lesser General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.
*/
#include "skypetab.h"
#include "settingsdialog.h"
#include <QTimer>
#include <QApplication>
#include <QMouseEvent>
#include <QPainter>
#include <QBrush>
#include <QPen>
#include <QColor>
#include "focusguard.h"
#include "mainwindow.h"
#include "qintercept.h"
#include <QtDebug>
#include <QX11Info>
#include "x11.h"
#include <X11/Xlib.h>
#include <stdio.h>
#include <dlfcn.h>
#include <queue>

namespace skypetab
{

SkypeTab::SkypeTab(QObject *parent) :
    QObject(parent)
{
	mainWindow=0;
	AddSignalIntercept("QSystemTrayIcon", SIGNAL(activated(QSystemTrayIcon::ActivationReason)),
			this, SLOT(onTrayMenuActivated(QSystemTrayIcon::ActivationReason)),
			SIGNAL(_raiseTrayMenuActivated(QSystemTrayIcon::ActivationReason)));
	//Make sure that this library won't be loaded into any other process
	QString env=QString::fromLocal8Bit(getenv("LD_PRELOAD"));
	Dl_info nfo;
	dladdr((void*)XCreateSimpleWindow, &nfo);
	QString path=QString::fromLocal8Bit(nfo.dli_fname);
	env=env.remove(path).remove("libskypetab-ng.so");
	setenv("LD_PRELOAD", env.toLocal8Bit().data(), 1);

}

void SkypeTab::_stage2Init()
{
	if(mainWindow!=0)
		return;
	mainWindow=new STabMainWindow();
	_trayIcon=0;
	_trayMenu=0;
	_myMenu=new QMenu("SkypeTab");
	_myMenu->addAction("Settings", new SettingsDialog(), SLOT(execIt()));
	_myMenu->addAction("About SkypeTab", new AboutDialog(), SLOT(exec()));


	printf("Created main window\n");

	if(settings.value("x11/swallowErrors", QVariant::fromValue(false)).toBool())
		X11::SetupErrorHandler();

	this->startTimer(50);
	_stagingArea=(new QWidget())->winId();
}


void SkypeTab::stage0Init()
{
	static bool done=false;
	if(done)
		return;
	done=true;
	foreach (QString cmdArg, QApplication::instance()->arguments())
	{
		QString wClPrefix = "--skypetab-class=";
		if(cmdArg.startsWith(wClPrefix))
		{
			SkypeTab::windowClassName=cmdArg.mid(wClPrefix.length());
		}
	}

	if(STabMainWindow::tryActivatePreviousInstance())
		exit(0);
}

void SkypeTab::stage1Init()
{
	static bool done=false;
	if(done)
		return;
	done=true;
	X11::XDisplay=QX11Info::display();
	stage0Init();
	const char*curDesktop= getenv("XDG_CURRENT_DESKTOP");
	if(curDesktop!=0)
	{
		if(0==strcmp("Unity", curDesktop))
			winManager=Unity;
	}

	_instance=new SkypeTab();
}

void SkypeTab::stage2Init()
{
	static bool done=false;
	if(done)
		return;
	done=true;
	stage1Init();
	_instance->_stage2Init();
}

bool SkypeTab::needTabActivation (QWidget* widget)
{
	QString cl=QString::fromLocal8Bit(widget->metaObject()->className());
	if((settings.value("tabs/disableTransferActivation", QVariant::fromValue(false)).toBool()) && cl=="Skype::TransferWindow")
		return false;
	return true;
}

bool SkypeTab::onWindowActivation(QWidget *widget)
{
	stage2Init();
	while(widget->parentWidget())
		widget=widget->parentWidget();
	QString cl=QString::fromLocal8Bit(widget->metaObject()->className());
	if(!needTabActivation(widget))
		return false;
	if(_instance->mainWindow->activateTab(widget))
	{
		QApplication::setActiveWindow(widget);
		X11::Flush();
		X11::Sync(false);
		return false;
	}
	return true;
}


QIcon SkypeTab::onSetIcon(const QIcon& icon, QSystemTrayIcon* tray)
{
	QSize iconsize = icon.actualSize(QSize(99,99));
	if (_trayPixmap == 0)
		_trayPixmap = new QPixmap(iconsize);

	if (_trayIcon == 0 && tray != 0)
		_trayIcon = tray;
	*_trayPixmap = icon.pixmap(iconsize);
	return GetIcon();

}

QIcon SkypeTab::GetIcon()
{
	QPixmap newPixmap = *_trayPixmap;
	QPainter painter(&newPixmap);

	QString string = "";
	if (_activeMsgsCount == 0)
		return QIcon(newPixmap);
	if(_activeMsgsCount<10)
		string = QString::number(_activeMsgsCount);
	else
		string = tr("M");
	painter.setRenderHints(QPainter::Antialiasing);

	QBrush back(QColor(40, 40, 120, 255));
	QPen front(QColor(255, 255, 255, 255));
	painter.setBrush(back);
	painter.drawEllipse(QPoint(newPixmap.width()-7, 7), 7, 7);

	painter.setPen(front);
	QFont font=painter.font();

	font.setBold(true);
	font.setPointSize(8);
	painter.setFont(font);
	painter.drawText(newPixmap.width()-15, 1, 14, 14, Qt::AlignHCenter | Qt::AlignVCenter, string);

	return QIcon(newPixmap);
}

void SkypeTab::updateTrayIcon(int count)
{
	if(_activeMsgsCount==count)
		return;
	_activeMsgsCount = count;
	if (_trayPixmap==0)
		return;
	if (_trayIcon != 0) {
		(_trayIcon->*realSetIcon)(GetIcon());
	}
}


void SkypeTab::onMenuShow()
{
	if(_trayMenu==0)
		return;
	QList<QAction*> actions= _trayMenu->actions();
	if(actions.length()==0)
		return;

	_trayMenu->insertMenu(actions[actions.length()-1],_myMenu);

}

void skypetab::SkypeTab::onSetContextMenu(QMenu *menu)
{
	stage2Init();
	if(_instance->_trayMenu!=0)
		return;
	_instance->_trayMenu=menu;
	connect(menu, SIGNAL(aboutToShow()),_instance, SLOT(onMenuShow()));
	_instance->onMenuShow();
}

bool SkypeTab::mainSkypeWindowEnabled()
{
	if(!_instance)
		return false;
	if(!_mainSkypeWindow)
		return false;
	if(!_instance->_newStyleContactList)
		return true;
	if(_mainSkypeWindow->windowTitle().startsWith("Skype™"))
		return false;
	return true;
}

void SkypeTab::onTrayMenuActivated(QSystemTrayIcon::ActivationReason reason)
{
	stage2Init();
	if(reason!=QSystemTrayIcon::Context && mainSkypeWindowEnabled())
	{
		onTrayIcon();
	}
	else
		raiseTrayMenuActivated(reason);

}

void SkypeTab::onMainWindowDestroyed()
{
	_mainSkypeWindow = 0;
}



SkypeTab* SkypeTab::_instance=0;
QWidget*SkypeTab::_mainSkypeWindow=0;
QPixmap*SkypeTab::_trayPixmap=0;
QSystemTrayIcon*SkypeTab::_trayIcon=0;
int SkypeTab::_activeMsgsCount=0;
bool SkypeTab::_manualTrayUpdate=false;
QSettings SkypeTab::settings("kekekeks","skypetab-ng");
WinManager SkypeTab::winManager=Generic;
bool* SkypeTab::enabledTabClassesList=0;
QString SkypeTab::windowClassName = "SkypeTab";

const char* SkypeTab::tabClassesList[][2]={
	{"Skype::ChatWindow", "Chats"},
	{"Skype::TransferWindow", "File transfers"},
	{"",""},//default classes are above this line
	{"Skype::CallWindow", "Calls"},
	{"Skype::CallView", "Skype 4.0 Calls"}, //FIXME: Add some code to merge that with previous line
{0,0}};



WId SkypeTab::onNewWindow()
{
	//We should be completely initialized at this moment

	QWidget*widget=WindowCreationInitiator;
	if(widget==0)
		return 0;
	const QMetaObject* meta=widget->metaObject();


	loadEnabledTabClassesList();
	const char* classname=meta->className();
	printf("Creating window for %s\n", classname);

	for(int c=0; tabClassesList[c][0]!=0; c++)
	{
		const char *name=tabClassesList[c][0];
		if(0==strcmp(classname, name))
		{
			QString keyName("tabClasses/");
			keyName=keyName.append(classname);

			if(enabledTabClassesList[c])
			{
				printf("%s is under control", classname);
				_instance->_pendingWindows.append(widget);
				return _instance->_stagingArea;
			}
			break;
		}
	}
	return 0;

}

void SkypeTab::onTryShow(QWidget *widget)
{
	if (_mainSkypeWindow)
		return;
	stage2Init();

	//Walk on widget's hierarhy
	while (widget)
	{
		const char *className = widget->metaObject()->className();
		if(0==strcmp(className, "QWidget") || 0==strcmp(className, "QMainWindow"))
		{
			QString title=widget->windowTitle();
			QString clName=widget->metaObject()->className();
			if(clName=="QWidget" || clName == "QMainWindow")
			{
				bool oldSkype = title.contains("Skype") && (title.contains("Beta") || title.contains("4.0"));
				bool newSkype = title.contains("- Skype™")||clName =="QMainWindow";
				if(oldSkype || newSkype)
				{
					if(!oldSkype)
						_instance->_newStyleContactList=true;
					_mainSkypeWindow=widget;
					connect(widget, SIGNAL(destroyed()), _instance, SLOT(onMainWindowDestroyed()));
					_instance->mainWindow->SetMainWindow(widget);
					_instance->_oldMainWindowEnabled = true;
					if(!settings.value("startup/hidden", QVariant::fromValue(false)).toBool())
						_instance->mainWindow->show();
					break;
				}
			}
		}
		widget=widget->parentWidget();
	}
}

void SkypeTab::raiseTrayMenuActivated(QSystemTrayIcon::ActivationReason reason)
{
	_raiseTrayMenuActivated(reason);
}

void SkypeTab::onTrayIcon()
{
	stage2Init();
	if((SkypeTab::winManager==Generic) && (mainWindow->isActiveWindow()||mainWindow->hasActiveTab()))
	{
		mainWindow->setWindowState(mainWindow->windowState() & ~Qt::WindowActive | Qt::WindowMinimized);
	}
	else
	{
		mainWindow->show();
		mainWindow->raise();
		mainWindow->activateWindow();
	}
}

void SkypeTab::timerEvent(QTimerEvent *)
{
	while(_pendingWindows.length()!=0)
	{
		QWidget*widget=_pendingWindows[0];
		_pendingWindows.removeAt(0);

		mainWindow->show();
		X11::Flush();

		bool activate = needTabActivation(widget);
		mainWindow->AddTab(widget, activate);
		if(activate)
			mainWindow->activateWindow();

	}
	bool mwEnabled=mainSkypeWindowEnabled();
	if(_oldMainWindowEnabled!=mwEnabled)
	{
		if(mwEnabled)
			mainWindow->show();
		else
			mainWindow->hide();
		_oldMainWindowEnabled=mwEnabled;
	}
}

void SkypeTab::loadEnabledTabClassesList()
{
	if(enabledTabClassesList!=0)
		return;
	int cnt=0;
	for(int c=0; tabClassesList[c][0]!=0; c++)
		cnt++;
	enabledTabClassesList=new bool[cnt];
	bool isDefault=true;
	for(int c=0; tabClassesList[c][0]!=0; c++)
	{
		const char *name=tabClassesList[c][0];
		QString keyName("tabClasses/");
		keyName=keyName.append(name);
		bool enabled= (
				((!settings.contains(keyName))&&isDefault) //If no setting for this entry, use default flag
				||
				(settings.value(keyName).toInt()==1)
		);
		enabledTabClassesList[c]=enabled;
		if(name[0]==0)
			isDefault=false;
	}
}
}

